### DISCUSSION TITLE
Fastapi recommends use gunicorn with uvicorn workers why?

### DISCUSSION THREAD

**Discussion by iamnotagentleman** [fastapi/fastapi#6419] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}
```

### Description

Currently i use uvicorn only in production but when i check the docs it says you should run uvicorn workers inside gunicorn, what do i get when i replace it with gunicorn?

### Operating System

Linux

### FastAPI Version

latest

### Python Version

3.9

**Comment by insomnes** [comment]
You can read exactly what you search in uvicorn documentation:
https://www.uvicorn.org/deployment/#using-a-process-manager

**Comment by adriangb** [comment]
I think it's important to look at the bigger picture in a production setup. If you're running containerized, replicated deployments on k8s, you don't really need a process manager for resiliency or upgrades. What you do need however is good load balancing, and Gunicorn / Uvicorn workers (well, really the OS itself) provide that. k8s load balancing is just iptables by default I believe, so not even round robin.

**Comment by iamnotagentleman** [comment]
@insomnes @adriangb thanks for answers,  we run our application on the k8s, but i still don't get it how a process manager effect the load balancing, pods are independent among themselves.

**Comment by insomnes** [comment]
There are recent updates in deployment section of docs, you can read about docker deployment and variants:
https://fastapi.tiangolo.com/deployment/docker/#deployment-concepts

**Comment by iamnotagentleman** [comment]
@insomnes thanks a lot

**Comment by adriangb** [comment, marked as answer]
> i still don't get it how a process manager effect the load balancing, pods are independent among themselves

k8s picks a pod at random (someone please correct me if I'm wrong, as far as I know [it's just iptables rules w/ equal probability](https://stackoverflow.com/a/48796004), I don't think it's actual round robin). then if the pod is running Gunicorn / another process manager that uses a similar architecture, an idle worker will pick up the request (not a random worker, an idle worker).

So let's say your app takes 500ms to process each request.
If you just had 100 pods each running 1 process, there's a chance that if you get 2 concurrent requests they both end up at the same worker and so one of them takes 1000ms to process.
If instead you have 100 Gunicorn workers on 1 pod (not a good idea), you'd have to be processing 100 concurrent requests before that happens.
But there are advantages to having pods and not just running 1 pod with a 100 Gunicorn workers, so you probably want both.

Alternatively, I believe the best of both worlds is to use Envoy or another traffic aware reverse proxy / load balancer + a headless k8s Service , or a full blown service mesh architecture. But those are big architectural changes that come with a lot of added complexity, so it's not really a simple choice.
